File: src/test/java/org/jesperancinha/rockstarts/rockstarsmanager/controller/ArtistsControllerIT.java
package org.jesperancinha.rockstarts.rockstarsmanager.controller;

import org.jesperancinha.rockstarts.rockstarsmanager.containers.AbstractTestContainersIT;
import org.jesperancinha.rockstarts.rockstarsmanager.data.ArtistDto;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.transaction.annotation.Transactional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
import static org.springframework.test.annotation.DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD;

@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = RANDOM_PORT)
@ActiveProfiles("test")
@DirtiesContext(classMode = BEFORE_EACH_TEST_METHOD)
@ContextConfiguration(initializers = AbstractTestContainersIT.DockerPostgresDataInitializer.class)
@Transactional
public class ArtistsControllerIT {

    public static final String DUA_LIPA = "DuaLipa";
    public static final String ARIANA_GRANDE = "Ariana Grande";
    public static final String MABEL = "Mabel";

    @Autowired
    private ArtistsController controller;

    @Test
    void givenArtistOnDb_whenGettingArtist_thenReturnArtistNoError() {
        ResponseEntity<ArtistDto> artistsById = this.controller.getArtistsById(123L);

        assertThat(artistsById).isNotNull();
        assertThat(artistsById.getBody()).isNull();
        assertThat(artistsById.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }

    @Test
    void givenArtistOnDb_whenGetByName_thenGetArtistByName() {
        final ArtistDto artistDto = ArtistDto.builder()
                .id(10101L)
                .name(DUA_LIPA)
                .build();
        final ArtistDto savedToDataBase = this.controller.saveArtist(artistDto);
        ResponseEntity<ArtistDto> artistsByName = this.controller.getArtistByName(DUA_LIPA);

        assertThat(savedToDataBase).isNotNull();
        assertThat(savedToDataBase.getName()).isEqualTo(artistDto.getName());
        assertThat(artistsByName).isNotNull();
        assertThat(artistsByName.getStatusCode()).isEqualTo(HttpStatus.OK);
        final ArtistDto artistsByNameBody = artistsByName.getBody();
        assertThat(artistsByNameBody).isNotNull();
        assertThat(artistsByNameBody.getName()).isEqualTo(artistDto.getName());
    }

    @Test
    void givenArtist_whenSave_thenSaveArtist() {
        final ArtistDto artistDto = ArtistDto.builder()
                .id(10102L)
                .name(ARIANA_GRANDE)
                .build();

        final ArtistDto savedToDataBase = this.controller.saveArtist(artistDto);

        assertThat(savedToDataBase.getName()).isEqualTo(artistDto.getName());
        assertThat(savedToDataBase.getId()).isEqualTo(1L);
    }

    @Test
    void given2Artists_whenSave_thenSave2Artists() {
        final ArtistDto artistDtoAriana = ArtistDto.builder()
                .id(10102L)
                .name(ARIANA_GRANDE)
                .build();
        final ArtistDto artistDtoMabel = ArtistDto.builder()
                .id(10101L)
                .name(MABEL)
                .build();

        final ArtistDto savedAriana = this.controller.saveArtist(artistDtoAriana);
        final ArtistDto savedMabel = this.controller.saveArtist(artistDtoMabel);

        assertThat(savedAriana.getName()).isEqualTo(artistDtoAriana.getName());
        assertThat(savedMabel.getName()).isEqualTo(artistDtoMabel.getName());
        assertThat(savedAriana.getId()).isNotEqualTo(savedMabel.getId());
    }

    @Test
    void givenArtistOnDb_whenUpdate_thenUpdate() {
        final ArtistDto artistDto = ArtistDto.builder()
                .name(ARIANA_GRANDE)
                .build();
        final ArtistDto savedToDataBase = this.controller.saveArtist(artistDto);
        final ArtistDto artistDtoUpdate = ArtistDto.builder()
                .id(savedToDataBase.getId())
                .name(MABEL)
                .build();

        ArtistDto artistDtoUpdateResult = this.controller.putArtist(artistDtoUpdate, artistDtoUpdate.getId());

        assertThat(savedToDataBase).isNotNull();
        assertThat(savedToDataBase.getName()).isEqualTo(artistDto.getName());
        assertThat(artistDtoUpdate).isNotNull();
        assertThat(artistDtoUpdateResult).isNotNull();
        assertThat(artistDtoUpdateResult.getName()).isEqualTo(artistDtoUpdate.getName());
        assertThat(artistDtoUpdateResult.getId()).isEqualTo(savedToDataBase.getId());
    }

    @Test
    void givenArtistOnDb_whenDelete_thenCallsDelete() {
        final ArtistDto artistDto = ArtistDto.builder()
                .id(10102L)
                .name(ARIANA_GRANDE)
                .build();
        final ArtistDto savedToDataBase = this.controller.saveArtist(artistDto);

        this.controller.deleteById(savedToDataBase.getId());
        final ResponseEntity<ArtistDto> artistsById = this.controller.getArtistsById(savedToDataBase.getId());

        assertThat(savedToDataBase.getName()).isEqualTo(artistDto.getName());
        assertThat(artistsById).isNotNull();
        assertThat(artistsById.getBody()).isNull();
        assertThat(artistsById.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }
}

File: src/test/kotlin/org/jesperancinha/rockstarts/rockstarsmanager/controller/ArtistsControllerIT.kt
package org.jesperancinha.rockstarts.rockstarsmanager.controller

import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNot
import io.kotest.matchers.shouldNotBe
import org.assertj.core.api.Assertions
import org.jesperancinha.rockstarts.rockstarsmanager.containers.AbstractTestContainersIT.DockerPostgresDataInitializer
import org.jesperancinha.rockstarts.rockstarsmanager.data.ArtistDto
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.*
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.annotation.DirtiesContext.ClassMode
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.transaction.annotation.Transactional

@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@ActiveProfiles("test")
@ContextConfiguration(initializers = [DockerPostgresDataInitializer::class])
@Transactional
class ArtistsControllerIT @Autowired constructor(
    @Autowired
    private val controller: ArtistsController,
) {

    @Test
    fun givenArtistOnDb_whenGettingArtist_thenReturnArtistNoError() {
        val artistsById = controller.getArtistsById(123L)
        Assertions.assertThat(artistsById).isNotNull
        Assertions.assertThat(artistsById.body).isNull()
        Assertions.assertThat(artistsById.statusCode).isEqualTo(NOT_FOUND)
    }

    @Test
    fun givenArtistOnDb_whenGetByName_thenGetArtistByName() {
        val artistDto = ArtistDto(
            id = 10101L,
            name = DUA_LIPA
        )
        val savedToDataBase = controller.saveArtist(artistDto)
        val artistsByName = controller.getArtistByName(DUA_LIPA)
        savedToDataBase.shouldNotBeNull()
        savedToDataBase.name shouldBe artistDto.name
        artistsByName.shouldNotBeNull()
        artistsByName.statusCode shouldBe OK
        val artistsByNameBody = artistsByName.body
        artistsByNameBody.shouldNotBeNull()
        artistsByNameBody.name shouldBe artistDto.name
    }

    @Test
    fun givenArtist_whenSave_thenSaveArtist() {
        val artistDto = ArtistDto(
            id = 10102L,
            name = ARIANA_GRANDE)
        val savedToDataBase = controller.saveArtist(artistDto)
        savedToDataBase.shouldNotBeNull()
        savedToDataBase.name shouldBe artistDto.name
        savedToDataBase.id shouldBe 1L
    }

    @Test
    fun given2Artists_whenSave_thenSave2Artists() {
        val artistDtoAriana = ArtistDto(
            id =10102L,
            name = ARIANA_GRANDE)
        val artistDtoMabel = ArtistDto(
            id =10101L,
            name =MABEL)
        val savedAriana = controller.saveArtist(artistDtoAriana)
        val savedMabel = controller.saveArtist(artistDtoMabel)
        savedAriana.name shouldBe artistDtoAriana.name
        savedMabel.name shouldBe artistDtoMabel.name
        savedAriana.name shouldNotBe savedMabel.name
    }

    @Test
    fun givenArtistOnDb_whenUpdate_thenUpdate() {
        val artistDto = ArtistDto(
             name = ARIANA_GRANDE)
        val savedToDataBase = controller.saveArtist(artistDto)
        val artistDtoUpdate = ArtistDto(
             id = savedToDataBase.id,
             name = MABEL)
        val artistDtoUpdateResult = controller.putArtist(artistDtoUpdate, artistDtoUpdate.id)
        savedToDataBase.shouldNotBeNull()
        savedToDataBase.name shouldBe artistDto.name
        artistDtoUpdate.shouldNotBeNull()
        artistDtoUpdateResult.shouldNotBeNull()
        artistDtoUpdateResult.name shouldBe artistDtoUpdate.name
        artistDtoUpdateResult.id shouldBe savedToDataBase.id
    }

    @Test
    fun givenArtistOnDb_whenDelete_thenCallsDelete() {
        val artistDto = ArtistDto(
            id = 10102L,
            name= ARIANA_GRANDE)
        val savedToDataBase = controller.saveArtist(artistDto)
        savedToDataBase.id?.let { controller.deleteById(it) }
        val artistsById = savedToDataBase.id?.let { controller.getArtistsById(it) }
        savedToDataBase.name shouldBe artistDto.name
        artistsById.shouldNotBeNull()
        artistsById.body.shouldBeNull()
        artistsById.statusCode shouldBe NOT_FOUND
    }

    companion object {
        const val DUA_LIPA = "DuaLipa"
        const val ARIANA_GRANDE = "Ariana Grande"
        const val MABEL = "Mabel"
    }
}